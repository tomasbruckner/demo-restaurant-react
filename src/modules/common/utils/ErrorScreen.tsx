import React, { PureComponent } from 'react';
import PageLayout from '../layouts/PageLayout';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';

export class ErrorScreen extends PureComponent<{ title: string } & WithTranslation> {
  render() {
    const { t } = this.props;

    return <PageLayout title={t('page.error.title')} />;
  }
}

export default withTranslation(I18N_CORE)(ErrorScreen);
