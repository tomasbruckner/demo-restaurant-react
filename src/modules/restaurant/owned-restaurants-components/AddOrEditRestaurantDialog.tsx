import React, { PureComponent } from 'react';
import Button from 'modules/common/components/controls/Button';
import AddCircleOutline from '@material-ui/icons/AddCircleOutline';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import Input from 'modules/common/components/controls/Input';
import FormDialog from 'modules/common/components/dialogs/FormDialog';
import { withStyles } from '@material-ui/core';
import { connect, ConnectedProps } from 'react-redux';
import { restaurantCreateAsync, restaurantEditAsync } from 'core/redux/ducks/RestaurantDuck';
import { RouteComponentProps, withRouter } from 'react-router';
import { getRestaurantDetail } from 'core/router/urls';
import { RestaurantResponse } from 'services/restaurantService';
import { setErrorNotification } from 'core/redux/ducks/NotificationDuck';

const MyInput = withStyles(theme => ({
  root: {
    margin: theme.spacing(1),
    width: '98%',
  },
}))(Input);

const connector = connect(null, {
  createRestaurant: restaurantCreateAsync,
  editRestaurant: restaurantEditAsync,
  dispatchErrorNotification: setErrorNotification,
});

type Props = {
  defaultRestaurantValues?: RestaurantResponse;
  hideButton?: boolean;
  onOpenToggle?: () => void;
  isControlledOpen?: boolean;
};

export class AddOrEditRestaurantDialog extends PureComponent<
  Props & RouteComponentProps & ConnectedProps<typeof connector> & WithTranslation
> {
  state = {
    isOpen: false,
  };

  onOpenToggle = () => {
    const { isOpen } = this.state;
    const { onOpenToggle } = this.props;

    if (onOpenToggle) {
      onOpenToggle();
    }
    this.setState({ isOpen: !isOpen });
  };

  onSubmit = async (data: any = {}) => {
    const { name, phone, description, address, url } = data.target.elements;
    const {
      createRestaurant,
      editRestaurant,
      history,
      defaultRestaurantValues,
      dispatchErrorNotification,
      t,
    } = this.props;
    const requestData = {
      address: address.value,
      name: name.value,
      contactPhone: phone.value,
      description: description.value,
      websiteUrl: url.value,
    };

    if (defaultRestaurantValues) {
      const { restaurantId } = defaultRestaurantValues;
      try {
        await editRestaurant({ restaurantId, data: requestData });
      } catch (e) {
        dispatchErrorNotification({ message: t('component.restaurant-add-dialog.error-conflict') });
      }
      this.onOpenToggle();
    } else {
      try {
        const res = await createRestaurant(requestData);
        const { restaurantId } = res.payload as RestaurantResponse;
        history.push(getRestaurantDetail(restaurantId));
      } catch (e) {
        dispatchErrorNotification({ message: t('component.restaurant-add-dialog.error-conflict') });
      }
    }
  };

  render() {
    const { t, defaultRestaurantValues, isControlledOpen, hideButton } = this.props;
    const { isOpen } = this.state;

    return (
      <>
        {hideButton ? null : (
          <Button
            variant="outlined"
            color="primary"
            data-cy="restaurant-create-button"
            startIcon={<AddCircleOutline />}
            onClick={this.onOpenToggle}
          >
            {t('page.owned-restaurants.buttons.add')}
          </Button>
        )}
        <FormDialog
          title={t('component.restaurant-add-dialog.title')}
          submitButtonText={t('component.restaurant-add-dialog.submit')}
          cancelButtonText={t('component.restaurant-add-dialog.cancel')}
          onClose={this.onOpenToggle}
          isOpen={isControlledOpen !== undefined ? isControlledOpen : isOpen}
          onSubmit={this.onSubmit}
        >
          <MyInput
            id="idReviewReplyDialogName"
            label={t('component.restaurant-add-dialog.name')}
            name="name"
            autoFocus
            required
            data-cy="restaurant-add-name"
            defaultValue={defaultRestaurantValues ? defaultRestaurantValues.name : undefined}
            inputProps={{
              maxLength: 500,
            }}
            fullWidth
          />
          <MyInput
            id="idReviewReplyDialogWebsiteUrl"
            label={t('component.restaurant-add-dialog.website-url')}
            name="url"
            required
            data-cy="restaurant-add-url"
            defaultValue={defaultRestaurantValues ? defaultRestaurantValues.websiteUrl : undefined}
            inputProps={{
              maxLength: 500,
            }}
            fullWidth
          />
          <MyInput
            id="idReviewReplyDialogContactPhone"
            label={t('component.restaurant-add-dialog.phone')}
            name="phone"
            required
            data-cy="restaurant-add-phone"
            defaultValue={
              defaultRestaurantValues ? defaultRestaurantValues.contactPhone : undefined
            }
            inputProps={{
              maxLength: 100,
            }}
            fullWidth
          />
          <MyInput
            id="idReviewReplyDialogAddress"
            label={t('component.restaurant-add-dialog.address')}
            name="address"
            required
            data-cy="restaurant-add-address"
            defaultValue={defaultRestaurantValues ? defaultRestaurantValues.address : undefined}
            inputProps={{
              maxLength: 500,
            }}
            fullWidth
          />
          <MyInput
            id="idReviewReplyDialogDescription"
            label={t('component.restaurant-add-dialog.description')}
            name="description"
            required
            data-cy="restaurant-add-description"
            defaultValue={defaultRestaurantValues ? defaultRestaurantValues.description : undefined}
            fullWidth
            rows={8}
            multiline
          />
        </FormDialog>
      </>
    );
  }
}

export default withTranslation(I18N_CORE)(connector(withRouter(AddOrEditRestaurantDialog)));
