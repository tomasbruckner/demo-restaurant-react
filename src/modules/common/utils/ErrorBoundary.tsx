import React, { PureComponent } from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';
import Logger from 'utils/logger';
import { I18N_CORE } from 'core/i18n/i18n';
import ErrorScreen from './ErrorScreen';

export class ErrorBoundaryComp extends PureComponent<
  { renderProp?: () => React.ReactNode } & WithTranslation
> {
  state = {
    error: null,
  };

  static getDerivedStateFromError(error: Error) {
    return { error };
  }

  componentDidCatch(error: unknown, errorInfo: unknown) {
    Logger.error(error, errorInfo);
  }

  render() {
    const { children, renderProp, t } = this.props;
    const { error } = this.state;

    if (!error) {
      return children;
    }

    if (renderProp) {
      return renderProp();
    }

    return <ErrorScreen title={t('CommonComponents.ErrorBoundary.error')} />;
  }
}

export default withTranslation(I18N_CORE)(ErrorBoundaryComp);
