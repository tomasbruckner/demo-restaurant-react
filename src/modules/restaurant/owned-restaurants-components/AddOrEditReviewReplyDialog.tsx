import React, { PureComponent } from 'react';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import Button from 'modules/common/components/controls/Button';
import FormDialog from 'modules/common/components/dialogs/FormDialog';
import Input from 'modules/common/components/controls/Input';
import styled from 'styled-components/macro';
import { connect, ConnectedProps } from 'react-redux';
import { getPendingReviewsAsync, reviewFilterAsync } from 'core/redux/ducks/ReviewDuck';
import { reviewReplyCreateAsync, reviewReplyEditAsync } from 'core/redux/ducks/ReviewReplyDuck';
import { DEFAULT_PAGE_SIZE } from '../../../utils/constants';

const StyledButton = styled(Button)`
  margin: 32px;
`;

const connector = connect(null, {
  dispatchGetPendingReviews: getPendingReviewsAsync,
  dispatchGetReviews: reviewFilterAsync,
  dispatchCreateReply: reviewReplyCreateAsync,
  dispatchEditReply: reviewReplyEditAsync,
});

type Props = {
  restaurantId: number;
  reviewId: number;
  reviewReplyId?: number;
  defaultOwnerResponse?: string;
  hideButton?: boolean;
  isControlledOpen?: boolean;
  onToggle?: () => void;
};

export class AddOrEditReviewReplyDialog extends PureComponent<
  Props & ConnectedProps<typeof connector> & WithTranslation
> {
  state = {
    isOpen: false,
  };

  onOpenToggle = () => {
    const { isOpen } = this.state;
    const { onToggle } = this.props;

    if (onToggle) {
      onToggle();
    }

    this.setState({ isOpen: !isOpen });
  };

  onSubmit = async (data: any) => {
    const { text } = data.target.elements;
    const {
      dispatchGetPendingReviews,
      dispatchCreateReply,
      dispatchEditReply,
      dispatchGetReviews,
      reviewReplyId,
      restaurantId,
      reviewId,
    } = this.props;

    if (reviewReplyId) {
      await dispatchEditReply({
        data: {
          text: text.value,
        },
        reviewReplyId,
      });
      dispatchGetReviews({
        sort: {
          predicate: 'created',
          reverse: true,
        },
        skip: 0,
        take: DEFAULT_PAGE_SIZE,
        filter: {
          restaurantId,
        },
      });
    } else {
      await dispatchCreateReply({
        text: text.value,
        reviewId,
      });
      dispatchGetPendingReviews(restaurantId);
    }

    this.onOpenToggle();
  };

  render() {
    const { t, hideButton, defaultOwnerResponse, isControlledOpen } = this.props;
    const { isOpen } = this.state;

    return (
      <>
        {hideButton ? null : (
          <StyledButton
            color="primary"
            variant="outlined"
            onClick={this.onOpenToggle}
            data-cy="review-reply-dialog-open"
          >
            {t('component.review-pending-reply-dialog.click-reply')}
          </StyledButton>
        )}
        <FormDialog
          title={t('component.review-pending-reply-dialog.title')}
          submitButtonText={t('component.review-pending-reply-dialog.submit')}
          cancelButtonText={t('component.review-pending-reply-dialog.cancel')}
          onClose={this.onOpenToggle}
          isOpen={isControlledOpen !== undefined ? isControlledOpen : isOpen}
          onSubmit={this.onSubmit}
        >
          <Input
            id="idReviewReplyDialogText"
            label={t('component.review-pending-reply-dialog.reply')}
            autoFocus
            required
            data-cy="review-reply-dialog-text"
            name="text"
            defaultValue={defaultOwnerResponse}
            rows={8}
            multiline
            fullWidth
          />
        </FormDialog>
      </>
    );
  }
}

export default withTranslation(I18N_CORE)(connector(AddOrEditReviewReplyDialog));
