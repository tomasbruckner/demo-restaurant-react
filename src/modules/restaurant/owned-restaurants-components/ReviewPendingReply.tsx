import React, { PureComponent } from 'react';
import { Typography } from '@material-ui/core';
import RatingWrapper from '../common-components/RatingWrapper';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import { formatDate } from 'utils/commonUtils';
import Divider from 'modules/common/components/controls/Divider';
import ReviewReplyDialog from './AddOrEditReviewReplyDialog';
import { ReviewResponse } from 'services/reviewService';

export type Props = ReviewResponse;

export class ReviewPendingReply extends PureComponent<Props & WithTranslation> {
  render() {
    const { reviewRatingId, restaurantId, title, text, created, reviewId, t } = this.props;

    return (
      <div style={{ display: 'flex', marginBottom: '32px' }}>
        <Divider
          orientation="vertical"
          flexItem
          style={{ marginLeft: '32px', marginRight: '32px' }}
        />
        <div>
          <Typography gutterBottom variant="h4" component="h2">
            {title}
          </Typography>
          <RatingWrapper
            averageRating={reviewRatingId}
            canEdit={false}
            text={t('component.review-pending-reply.review-date', {
              reviewDate: formatDate(created),
            })}
          />
          <Typography variant="body2" component="p">
            {text}
          </Typography>
          <ReviewReplyDialog restaurantId={restaurantId} reviewId={reviewId} />
        </div>
      </div>
    );
  }
}

export default withTranslation(I18N_CORE)(ReviewPendingReply);
