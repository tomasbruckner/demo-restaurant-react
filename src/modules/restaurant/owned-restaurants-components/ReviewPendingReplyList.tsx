import React, { PureComponent } from 'react';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import ReviewPendingReply, { Props as PendingReplyProps } from './ReviewPendingReply';
import { PagedDto } from 'types/commonTypes';

export type Props = {
  reviewsPendingReply: PagedDto<PendingReplyProps>;
};

export class ReviewPendingReplyList extends PureComponent<Props & WithTranslation> {
  render() {
    const { reviewsPendingReply } = this.props;

    return reviewsPendingReply.data.map(o => <ReviewPendingReply key={o.reviewId} {...o} />);
  }
}

export default withTranslation(I18N_CORE)(ReviewPendingReplyList);
