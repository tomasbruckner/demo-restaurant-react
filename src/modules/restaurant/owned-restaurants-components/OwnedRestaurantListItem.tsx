import React, { PureComponent } from 'react';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardContent from '@material-ui/core/CardContent';
import { Typography } from '@material-ui/core';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import ReviewPendingReplyList from './ReviewPendingReplyList';
import { RouteComponentProps, withRouter } from 'react-router';
import { getRestaurantDetail } from 'core/router/urls';
import { RestaurantResponse } from 'services/restaurantService';
import RatingWrapper from '../common-components/RatingWrapper';
import { connect, ConnectedProps } from 'react-redux';
import { RootState } from 'core/redux/store';
import { getPendingReviews, getPendingReviewsAsync } from 'core/redux/ducks/ReviewDuck';

export type Props = RestaurantResponse;

const connector = connect(
  (state: RootState, props: Props) => {
    const { restaurantId } = props;

    return {
      pendingReviews: getPendingReviews(state, restaurantId),
    };
  },
  {
    dispatchGetPendingReviews: getPendingReviewsAsync,
  },
);

export class OwnedRestaurantListItem extends PureComponent<
  ConnectedProps<typeof connector> & Props & WithTranslation & RouteComponentProps
> {
  public componentDidMount() {
    const { dispatchGetPendingReviews, restaurantId } = this.props;

    dispatchGetPendingReviews(restaurantId);
  }

  redirect = () => {
    const { restaurantId, history } = this.props;

    history.push(getRestaurantDetail(restaurantId));
  };

  render() {
    const { name, numberOfReviews, averageRating, pendingReviews, t } = this.props;

    return (
      <Card style={{ margin: '32px' }}>
        <CardActionArea onClick={this.redirect}>
          <CardContent>
            <Typography gutterBottom variant="h4" component="h2">
              {name}
            </Typography>
            <RatingWrapper averageRating={averageRating} numberOfReviews={numberOfReviews} canEdit={false} />
          </CardContent>
        </CardActionArea>

        <Typography gutterBottom variant="body1" style={{ margin: '16px' }}>
          <i>
            {t(
              pendingReviews.size === 0
                ? 'component.owned-restaurant-item.no-pending'
                : 'component.owned-restaurant-item.pending',
            )}
          </i>
        </Typography>
        <ReviewPendingReplyList reviewsPendingReply={pendingReviews} />
      </Card>
    );
  }
}

export default withTranslation(I18N_CORE)(withRouter(connector(OwnedRestaurantListItem)));
