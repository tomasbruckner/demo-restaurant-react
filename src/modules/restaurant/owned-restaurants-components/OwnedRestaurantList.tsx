import React, { PureComponent } from 'react';
import OwnedRestaurantListItem, { Props as RestaurantProps } from './OwnedRestaurantListItem';
import { PagedDto } from 'types/commonTypes';

type Props = {
  restaurants: PagedDto<RestaurantProps>;
};

export class OwnedRestaurantList extends PureComponent<Props> {
  render() {
    const { restaurants } = this.props;

    return restaurants.data.map(props => (
      <OwnedRestaurantListItem key={props.restaurantId} {...props} />
    ));
  }
}

export default OwnedRestaurantList;
