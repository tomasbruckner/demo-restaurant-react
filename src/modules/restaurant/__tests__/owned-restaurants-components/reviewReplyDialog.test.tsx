import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { AddOrEditReviewReplyDialog } from 'modules/restaurant/owned-restaurants-components/AddOrEditReviewReplyDialog';
import { fakeFunction, fakeTranslationProps } from 'utils/testUtils';

describe('ReviewReplyDialog', () => {
  it('basic example', () => {
    const tree = renderer
      .create(
        <AddOrEditReviewReplyDialog
          reviewId={1}
          restaurantId={2}
          dispatchGetPendingReviews={fakeFunction}
          dispatchCreateReply={fakeFunction}
          {...fakeTranslationProps}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
