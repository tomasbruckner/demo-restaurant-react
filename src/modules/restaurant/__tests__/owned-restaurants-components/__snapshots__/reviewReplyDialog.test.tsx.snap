// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewReplyDialog basic example 1`] = `
.c0 {
  margin: 32px;
}

<button
  className="MuiButtonBase-root MuiButton-root MuiButton-outlined c0 MuiButton-outlinedPrimary"
  data-cy="review-reply-dialog-open"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    component.review-pending-reply-dialog.click-reply
  </span>
</button>
`;
